vignettes/contribution.Rmd

```diff
@@ -0,0 +1,371 @@
+
+## Github conventions
+
+When using github to collaborate, the following conversions are needed
+
+1. Github issues is for issues, feature requests, bugs. Before creating a issue, please make sure this issue has not been reported yet.
+    * If you are going to work on this issue, please assign yourself.
+2. Please create a branch in the `mmrm` repository, instead of creating forks, unless you are not yet a team member.
+    * The name of a branch should be like `<issue_id>_<short_discription>`. Use underscore to separate the description.
+3. Please create a Pull Request when you think your code changes are ready:
+    * they are well documented
+    * they have corresponding unit tests.
+    * they pass all the github action checks
+4. Address all the comments you receive.
+    * at least one approval is needed to merge.
+
+# How to contribute?
+
+## How to set up development environment?
+
+The development of `mmrm` package is based on latest R and C++ compilers.
+The package dependencies are most recent versions from CRAN.
+We recommend that your working environment is set up in the same way.
+But there are some tools we recommend you to install:
+
+1. `docker`. This tool will create separated environment which allows you to separate the host system and development system.
+2. `gitkraken`. This tool is a super useful in visualizing the git commit graphs, file history, etc.
+3. `lintr`. This R package will allow you to do the static code analysis. See [.lintr](#.lintr)
+4. `pre-commit`. This is a python module that allow you to identify issues before you commit. See [.pre-commit-config.yaml](#.pre-commit-config.yaml)
+
+## How to pick issues?
+
+The issues are categorized with several labels:
+
+1. SPx. SP means complexity, the larger the number, the more difficult the issue is.
+2. priority. Issues with this label should be completed with higher priority.
+3. good first issue. Issues with this label are good choices for new team members.
+
+Other labels are straightforward.
+Please choose issue based on your interest, issue complexity and priority.
+
+## How to add a new test case?
+
+To add a new test case, you need to first identify the test scope.
+Does the test fit in the scope of existing tests?
+If so, please modify the existing test files under `tests/testthat` folder or `src` folder, depending on the code to be tested is R or C++.
+Otherwise please create a new test file, with its name prefix as `test-`.
+
+In each test case, use the following structure.
+
+```{r eval=FALSE}
+test_that("<function_name> do something as explained", {
+  # test body
+  expect_identical(1, 1)
+})
+```
+
+The purpose of the test should be clearly stated first.
+
+In the test body part, conduct the tests, e.g. use `expect_identical` to check consistency, `expect_error` to catch error messages, etc.
+The test body should not follow the same implementation logic as the package did, otherwise we may fall into the same issue.
```

Comment:
```suggestion
The test body should not follow the same implementation logic as the package did, otherwise we may miss mistakes in implementation.
```